File: app/src/main/java/com/salesquest/sales_quest/ui/customers/CustomerFormPage.kt
package com.salesquest.sales_quest.ui.customers

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Note
import androidx.compose.material.icons.filled.Person
import androidx.compose.material.icons.filled.Phone
import androidx.compose.material3.CenterAlignedTopAppBar
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilterChip
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.salesquest.sales_quest.core.AppContainer
import com.salesquest.sales_quest.core.SaveCustomerParams
import com.salesquest.sales_quest.data.CustomerStage
import com.salesquest.sales_quest.data.Operator
import kotlinx.coroutines.launch

/** 新增/编辑客户页 - 所有字段可选 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CustomerFormPage(
    customerId: String?,
    onBack: () -> Unit = {},
    viewModel: CustomerFormViewModel = viewModel(factory = CustomerFormViewModel.factory(customerId))
) {
    val customer by viewModel.customer.collectAsState()
    val snackbarHostState = remember { SnackbarHostState() }
    val scope = rememberCoroutineScope()
    val isEdit = customerId != null

    var name by remember { mutableStateOf("") }
    var phone by remember { mutableStateOf("") }
    var note by remember { mutableStateOf("") }
    var packageName by remember { mutableStateOf("") }
    var actualCost by remember { mutableStateOf("") }
    var traffic by remember { mutableStateOf("") }
    var minutes by remember { mutableStateOf("") }
    var subCards by remember { mutableStateOf("") }
    var operator by remember { mutableStateOf(Operator.UNKNOWN) }
    var selfReportedCost by remember { mutableStateOf<Int?>(null) }
    var stage by remember { mutableStateOf(CustomerStage.NEW) }
    var broadband by remember { mutableStateOf(false) }
    var camera by remember { mutableStateOf(false) }
    var saving by remember { mutableStateOf(false) }

    LaunchedEffect(customer?.id) {
        val c = customer
        if (c != null) {
            name = c.name
            phone = c.phone
            note = c.note ?: ""
            packageName = c.packageName ?: ""
            actualCost = c.actualCost?.toString() ?: ""
            traffic = c.traffic ?: ""
            minutes = c.minutes ?: ""
            subCards = if (c.subCards > 0) c.subCards.toString() else ""
            operator = Operator.fromCode(c.operator)
            selfReportedCost = c.selfReportedCost
            stage = CustomerStage.fromCode(c.salesStage)
            broadband = c.broadband
            camera = c.camera
        }
    }

    fun save() {
        if (saving) return
        saving = true
        scope.launch {
            try {
                val params = SaveCustomerParams(
                    id = customerId,
                    name = name.trim().ifEmpty { null },
                    phone = phone.trim().ifEmpty { null },
                    operator = operator,
                    selfReportedCost = selfReportedCost,
                    actualCost = actualCost.trim().toIntOrNull(),
                    packageName = packageName.trim().ifEmpty { null },
                    traffic = traffic.trim().ifEmpty { null },
                    minutes = minutes.trim().ifEmpty { null },
                    broadband = broadband,
                    subCards = subCards.trim().toIntOrNull() ?: 0,
                    camera = camera,
                    stage = stage,
                    note = note.trim().ifEmpty { null }
                )
                val savedId = AppContainer.saveCustomer(params)
                snackbarHostState.showSnackbar("保存成功")
                if (isEdit && customerId != null) {
                    onBack()
                } else {
                    onBack()
                }
            } catch (e: Exception) {
                snackbarHostState.showSnackbar("保存失败: ${e.message}")
                saving = false
            }
        }
    }

    Scaffold(
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text(if (isEdit) "编辑客户" else "新增客户") },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(Icons.AutoMirrored.Filled.ArrowBack, contentDescription = "返回")
                    }
                },
                actions = {
                    TextButton(onClick = ::save, enabled = !saving) {
                        if (saving) {
                            CircularProgressIndicator(strokeWidth = 2.dp, modifier = Modifier.size(18.dp))
                        } else {
                            Text("保存")
                        }
                    }
                },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors()
            )
        },
        snackbarHost = { SnackbarHost(snackbarHostState) }
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 16.dp, vertical = 8.dp)
        ) {
            // === 基础信息 ===
            SectionTitle("基础信息")
            OutlinedTextField(
                value = name,
                onValueChange = { name = it },
                label = { Text("客户称呼") },
                placeholder = { Text("如: 张哥, 不填自动编号") },
                leadingIcon = { Icon(Icons.Filled.Person, contentDescription = null) },
                singleLine = true,
                modifier = Modifier.fillMaxWidth()
            )
            Spacer(Modifier.height(12.dp))
            OutlinedTextField(
                value = phone,
                onValueChange = { phone = it },
                label = { Text("手机号") },
                leadingIcon = { Icon(Icons.Filled.Phone, contentDescription = null) },
                singleLine = true,
                modifier = Modifier.fillMaxWidth()
            )
            Spacer(Modifier.height(12.dp))
            ChipGroup(
                label = "运营商",
                options = Operator.entries.map { it.label to it },
                selected = operator,
                onSelected = { operator = it }
            )
            Spacer(Modifier.height(12.dp))
            SelfCostChips(
                selected = selfReportedCost,
                onSelected = { selfReportedCost = it }
            )
            Spacer(Modifier.height(12.dp))
            OutlinedTextField(
                value = note,
                onValueChange = { note = it },
                label = { Text("备注") },
                placeholder = { Text("记录关键信息...") },
                leadingIcon = { Icon(Icons.Filled.Note, contentDescription = null) },
                minLines = 3,
                modifier = Modifier.fillMaxWidth()
            )

            // === 套餐详情 (仅编辑时可填) ===
            if (isEdit) {
                Spacer(Modifier.height(24.dp))
                SectionTitle("套餐详情")
                OutlinedTextField(
                    value = packageName,
                    onValueChange = { packageName = it },
                    label = { Text("套餐名称") },
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(Modifier.height(12.dp))
                OutlinedTextField(
                    value = actualCost,
                    onValueChange = { actualCost = it },
                    label = { Text("实际月消费 (元)") },
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(Modifier.height(12.dp))
                OutlinedTextField(
                    value = traffic,
                    onValueChange = { traffic = it },
                    label = { Text("流量") },
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(Modifier.height(12.dp))
                OutlinedTextField(
                    value = minutes,
                    onValueChange = { minutes = it },
                    label = { Text("通话分钟") },
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth()
                )
                Spacer(Modifier.height(4.dp))
                SwitchRow("有宽带", broadband) { broadband = it }
                SwitchRow("有摄像头", camera) { camera = it }
                Spacer(Modifier.height(4.dp))
                OutlinedTextField(
                    value = subCards,
                    onValueChange = { subCards = it },
                    label = { Text("副卡数量") },
                    singleLine = true,
                    modifier = Modifier.fillMaxWidth()
                )
            }

            // === 客户状态 ===
            Spacer(Modifier.height(24.dp))
            SectionTitle("客户状态")
            ChipGroup(
                label = "当前状态",
                options = CustomerStage.entries.map { it.label to it },
                selected = stage,
                onSelected = { stage = it }
            )
            Spacer(Modifier.height(32.dp))
        }
    }
}

@Composable
fun SectionTitle(title: String) {
    Text(
        title,
        style = MaterialTheme.typography.titleSmall,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.padding(bottom = 8.dp)
    )
}

@Composable
fun <T> ChipGroup(
    label: String,
    options: List<Pair<String, T>>,
    selected: T,
    onSelected: (T) -> Unit
) {
    Column {
        Text(label, style = MaterialTheme.typography.labelMedium, modifier = Modifier.padding(start = 4.dp, bottom = 6.dp))
        androidx.compose.foundation.layout.FlowRow(
            horizontalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(8.dp)
        ) {
            options.forEach { (optionLabel, value) ->
                FilterChip(
                    selected = selected == value,
                    onClick = { onSelected(value) },
                    label = { Text(optionLabel) }
                )
            }
        }
    }
}

/** 月消费自报 ChoiceChip: 不清楚 / 60+ / 100+ / 150+ / 200+ / 300+ */
@Composable
fun SelfCostChips(
    selected: Int?,
    onSelected: (Int?) -> Unit
) {
    val options = listOf<Pair<String, Int?>>(
        "不清楚" to null,
        "60+" to 60,
        "100+" to 100,
        "150+" to 150,
        "200+" to 200,
        "300+" to 300
    )
    Column {
        Text("月消费自报", style = MaterialTheme.typography.labelMedium, modifier = Modifier.padding(start = 4.dp, bottom = 6.dp))
        androidx.compose.foundation.layout.FlowRow(
            horizontalArrangement = androidx.compose.foundation.layout.Arrangement.spacedBy(8.dp)
        ) {
            options.forEach { (optionLabel, value) ->
                FilterChip(
                    selected = selected == value,
                    onClick = { onSelected(value) },
                    label = { Text(optionLabel) }
                )
            }
        }
    }
}

@Composable
fun SwitchRow(label: String, checked: Boolean, onCheckedChange: (Boolean) -> Unit) {
    androidx.compose.material3.ListItem(
        headlineContent = { Text(label) },
        trailingContent = {
            Switch(checked = checked, onCheckedChange = onCheckedChange)
        }
    )
}

File: legacy/lib/ui/customers/customer_form_page.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import '../../data/database/app_database.dart';
import '../../models/enums.dart';
import '../../providers/customer_providers.dart';

/// 新增/编辑客户页 - 所有字段可选
class CustomerFormPage extends ConsumerStatefulWidget {
  final String? customerId;

  const CustomerFormPage({super.key, this.customerId});

  @override
  ConsumerState<CustomerFormPage> createState() => _CustomerFormPageState();
}

class _CustomerFormPageState extends ConsumerState<CustomerFormPage> {
  final _nameCtrl = TextEditingController();
  final _phoneCtrl = TextEditingController();
  final _noteCtrl = TextEditingController();
  final _packageCtrl = TextEditingController();
  final _actualCostCtrl = TextEditingController();
  final _trafficCtrl = TextEditingController();
  final _minutesCtrl = TextEditingController();
  final _subCardsCtrl = TextEditingController();

  Operator _operator = Operator.unknown;
  int? _selfReportedCost; // null = 不清楚
  CustomerStage _stage = CustomerStage.new_;
  bool _broadband = false;
  bool _camera = false;

  bool _initialized = false;
  bool _saving = false;

  bool get _isEdit => widget.customerId != null;

  @override
  void dispose() {
    _nameCtrl.dispose();
    _phoneCtrl.dispose();
    _noteCtrl.dispose();
    _packageCtrl.dispose();
    _actualCostCtrl.dispose();
    _trafficCtrl.dispose();
    _minutesCtrl.dispose();
    _subCardsCtrl.dispose();
    super.dispose();
  }

  /// 编辑时用已有数据预填表单
  void _prefill(CustomerEntity c) {
    _nameCtrl.text = c.name;
    _phoneCtrl.text = c.phone;
    _noteCtrl.text = c.note ?? '';
    _packageCtrl.text = c.packageName ?? '';
    _actualCostCtrl.text = c.actualCost?.toString() ?? '';
    _trafficCtrl.text = c.traffic ?? '';
    _minutesCtrl.text = c.minutes ?? '';
    _subCardsCtrl.text = c.subCards > 0 ? c.subCards.toString() : '';
    _operator = Operator.fromCode(c.operator);
    _selfReportedCost = c.selfReportedCost;
    _stage = CustomerStage.fromCode(c.salesStage);
    _broadband = c.broadband;
    _camera = c.camera;
  }

  Future<void> _save() async {
    if (_saving) return;
    setState(() => _saving = true);

    try {
      final params = SaveCustomerParams(
        id: widget.customerId,
        name: _nameCtrl.text.trim().isEmpty ? null : _nameCtrl.text.trim(),
        phone: _phoneCtrl.text.trim().isEmpty ? null : _phoneCtrl.text.trim(),
        operator: _operator,
        selfReportedCost: _selfReportedCost,
        actualCost: int.tryParse(_actualCostCtrl.text),
        packageName:
            _packageCtrl.text.trim().isEmpty ? null : _packageCtrl.text.trim(),
        traffic: _trafficCtrl.text.trim().isEmpty ? null : _trafficCtrl.text.trim(),
        minutes: _minutesCtrl.text.trim().isEmpty ? null : _minutesCtrl.text.trim(),
        broadband: _broadband,
        subCards: int.tryParse(_subCardsCtrl.text) ?? 0,
        camera: _camera,
        stage: _stage,
        note: _noteCtrl.text.trim().isEmpty ? null : _noteCtrl.text.trim(),
      );

      await ref.read(saveCustomerProvider(params).future);
      if (!mounted) return;
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('保存成功'), duration: Duration(seconds: 1)),
      );
      // 保存后返回上一页
      context.pop();
    } catch (e) {
      if (!mounted) return;
      ScaffoldMessenger.of(context).showSnackBar(SnackBar(content: Text('保存失败: $e')));
      setState(() => _saving = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final Widget body;
    if (_isEdit) {
      final detailAsync = ref.watch(customerDetailProvider(widget.customerId!));
      body = detailAsync.when(
        data: (customer) {
          if (customer == null) {
            return const Center(child: Text('客户不存在'));
          }
          if (!_initialized) {
            _prefill(customer);
            _initialized = true;
          }
          return _buildForm(context);
        },
        loading: () => const Center(child: CircularProgressIndicator()),
        error: (e, _) => Center(child: Text('加载失败: $e')),
      );
    } else {
      body = _buildForm(context);
    }

    return Scaffold(
      appBar: AppBar(
        title: Text(_isEdit ? '编辑客户' : '新增客户'),
        leading: IconButton(
          icon: const Icon(Icons.arrow_back),
          onPressed: () => context.pop(),
        ),
        actions: [
          TextButton(
            onPressed: _saving ? null : _save,
            child: _saving
                ? const SizedBox(
                    width: 18,
                    height: 18,
                    child: CircularProgressIndicator(strokeWidth: 2),
                  )
                : const Text('保存'),
          ),
        ],
      ),
      body: body,
    );
  }

  Widget _buildForm(BuildContext context) {
    return ListView(
      padding: const EdgeInsets.fromLTRB(16, 8, 16, 32),
      children: [
        // === 基础信息 ===
        _section(
          context,
          '基础信息',
          [
            TextField(
              controller: _nameCtrl,
              decoration: const InputDecoration(
                labelText: '客户称呼',
                hintText: '如: 张哥, 不填自动编号',
                prefixIcon: Icon(Icons.person_outline),
              ),
            ),
            const SizedBox(height: 12),
            TextField(
              controller: _phoneCtrl,
              decoration: const InputDecoration(
                labelText: '手机号',
                prefixIcon: Icon(Icons.phone_outlined),
              ),
              keyboardType: TextInputType.phone,
            ),
            const SizedBox(height: 12),
            _chipGroup(
              context: context,
              label: '运营商',
              options: Operator.values
                  .map((op) => (label: op.label, value: op))
                  .toList(),
              selected: _operator,
              onSelected: (v) => setState(() => _operator = v),
            ),
            const SizedBox(height: 12),
            _selfCostChips(context),
            const SizedBox(height: 12),
            TextField(
              controller: _noteCtrl,
              decoration: const InputDecoration(
                labelText: '备注',
                hintText: '记录关键信息...',
                prefixIcon: Icon(Icons.note_outlined),
              ),
              maxLines: 3,
            ),
          ],
        ),

        // === 套餐详情 (仅编辑时可填) ===
        if (_isEdit) ...[
          const SizedBox(height: 24),
          _section(
            context,
            '套餐详情',
            [
              TextField(
                controller: _packageCtrl,
                decoration: const InputDecoration(labelText: '套餐名称'),
              ),
              const SizedBox(height: 12),
              TextField(
                controller: _actualCostCtrl,
                decoration: const InputDecoration(labelText: '实际月消费 (元)'),
                keyboardType: TextInputType.number,
              ),
              const SizedBox(height: 12),
              TextField(
                controller: _trafficCtrl,
                decoration: const InputDecoration(labelText: '流量'),
              ),
              const SizedBox(height: 12),
              TextField(
                controller: _minutesCtrl,
                decoration: const InputDecoration(labelText: '通话分钟'),
              ),
              const SizedBox(height: 4),
              SwitchListTile(
                title: const Text('有宽带'),
                value: _broadband,
                onChanged: (v) => setState(() => _broadband = v),
              ),
              SwitchListTile(
                title: const Text('有摄像头'),
                value: _camera,
                onChanged: (v) => setState(() => _camera = v),
              ),
              const SizedBox(height: 4),
              TextField(
                controller: _subCardsCtrl,
                decoration: const InputDecoration(labelText: '副卡数量'),
                keyboardType: TextInputType.number,
              ),
            ],
          ),
        ],

        // === 客户状态 ===
        const SizedBox(height: 24),
        _section(
          context,
          '客户状态',
          [
            _chipGroup(
              context: context,
              label: '当前状态',
              options: CustomerStage.values
                  .map((s) => (label: s.label, value: s))
                  .toList(),
              selected: _stage,
              onSelected: (v) => setState(() => _stage = v),
            ),
          ],
        ),
      ],
    );
  }

  /// 分区: 标题 + 子项
  Widget _section(BuildContext context, String title, List<Widget> children) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(bottom: 8),
          child: Text(
            title,
            style: Theme.of(context)
                .textTheme
                .titleSmall
                ?.copyWith(fontWeight: FontWeight.bold),
          ),
        ),
        ...children,
      ],
    );
  }

  /// 通用 ChoiceChip 组
  Widget _chipGroup<T>({
    required BuildContext context,
    required String label,
    required List<({String label, T value})> options,
    required T selected,
    required ValueChanged<T> onSelected,
  }) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 4, bottom: 6),
          child: Text(label, style: Theme.of(context).textTheme.labelMedium),
        ),
        Wrap(
          spacing: 8,
          runSpacing: 4,
          children: options.map((o) {
            return ChoiceChip(
              label: Text(o.label),
              selected: selected == o.value,
              onSelected: (_) => onSelected(o.value),
            );
          }).toList(),
        ),
      ],
    );
  }

  /// 月消费自报 ChoiceChip: 不清楚 / 60+ / 100+ / 150+ / 200+ / 300+
  Widget _selfCostChips(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 4, bottom: 6),
          child: Text('月消费自报', style: Theme.of(context).textTheme.labelMedium),
        ),
        Wrap(
          spacing: 8,
          runSpacing: 4,
          children: _selfCostOptions.map((o) {
            return ChoiceChip(
              label: Text(o.label),
              selected: _selfReportedCost == o.value,
              onSelected: (_) => setState(() => _selfReportedCost = o.value),
            );
          }).toList(),
        ),
      ],
    );
  }

  static const List<({String label, int? value})> _selfCostOptions = [
    (label: '不清楚', value: null),
    (label: '60+', value: 60),
    (label: '100+', value: 100),
    (label: '150+', value: 150),
    (label: '200+', value: 200),
    (label: '300+', value: 300),
  ];
}
